accepted/2020/better-obsoletion/securestring-obsoletion.md

```diff
@@ -0,0 +1,190 @@
+
+One way to accomplish this is by using .NET 5.0's new [pinned object heap](https://github.com/dotnet/runtime/pull/32283). Arrays allocated on the heap are guaranteed not to be moved during a GC, so applications don't need to worry about sensitive contents being copied before the application logic has had a chance to clear the memory. This has the benefit that an application can treat the data like a normal array.
+
+Another way is to invest in [a proposed runtime feature](https://github.com/dotnet/runtime/issues/10480) which would make the GC more aggressive about zeroing memory on compaction. This has the benefit that _many_ different types – including standard strings – would become candidates for having their data zeroed out during a compaction phase. High-sensitivity applications may wish to opt in to this GC mode anyway depending on their risk analysis.
+
+### Minimizing misuse of sensitive data
+
+Here, the goal isn't to hide secrets from memory, but to signal to developers "this data is sensitive and should be handled cautiously." In other words, the goal is to minimize the risk of a developer _accidentally_ using the data in an insecure fashion, such as writing it to a log file. SecureString actually does a fine job of this by virtue of the fact that callers must drop down to unsafe code or use the `Marshal` class to interact with it.
+
+There is opportunity for .NET to provide a better type for this scenario, described in more detail later.
+
+### Using it because it has "secure" in the name
+
+Based both on internal and external conversations, there seems to be a significant audience using the SecureString type because it has "secure" in its name. I have seen many code bases which have contorted themselves into using SecureString in as many places as possible in order to fulfill of a well-intentioned but ill-executed "security compliance" edict. Since this sort of maneuvering generally involves unsafe code, this often ends up _introducing_ security bugs. (One partner was even using unsafe code to zero-out string instances after converting them to SecureString.)
+
+During these conversations it's apparent that they're not attempting to mitigate any specific threats. They're instead using SecureString because it's listed as a requirement in their engineering processes. These users would be best served by using the existing string class or by using the `NetworkCredential` wrapper as suggested earlier.
+
+## A proposal for shrouding data
+
+Of all the scenarios mentioned above, the one that strikes me as most interesting is the use of SecureString as a marker class to mean "you should be careful when using this data." It's possible to extend this into a more generalized proposal that can serve as a wrapper for arbitary types _T_, so it can be used to shroud both `byte` and `char`.
+
+```cs
+namespace System.Runtime.InteropServices
+{
+    public class ShroudedBuffer<T> : IDisposable where T : unmanaged
```

Comment:
As a non-native speaker I had to look up `shrouded` in the dictionary (and I've been consuming and writing English for the better part of my life) so yeah it's probably not the most obvious name 😄 

My suggestion would something like `HiddenBuffer<T>` or `SensitiveData<T>`.

Reply:
+1 to `Sensitive*<T>` - it's not hidden really, but it _is_ intended to convey meaning and "sensitive" conveys "be careful", like a fragile label on a package.

Reply:
`FoggyBuffer` or `LondonBuffer` 🤭😬

Reply:
FWIW in case its useful, Terraform uses the term `Sensitive` to describe the HCL flag for a similar purpose:
https://www.terraform.io/docs/extend/best-practices/sensitive-state.html#using-the-sensitive-flag